server/src/com/thoughtworks/go/server/service/EnvironmentConfigService.java

```diff
@@ -145,6 +136,24 @@ public EnvironmentConfig named(String environmentName) throws NoSuchEnvironmentE
         return environments.named(new CaseInsensitiveString(environmentName));
     }
 
+    public EnvironmentConfig getEnvironmentConfig(String environmentName) throws NoSuchEnvironmentException {
+        if(environments.hasEnvironmentNamed(new CaseInsensitiveString(environmentName)))
+            return environments.named(new CaseInsensitiveString(environmentName));
+        return null;
+    }
+
+    public void deleteEnvironment(String environmentName, HttpLocalizedOperationResult result) throws NoSuchEnvironmentException {
+        CruiseConfig config = goConfigService.getConfigForEditing();
+        EnvironmentsConfig environments = config.getEnvironments();
+        if (environments.remove(getEnvironmentConfig(environmentName))) {
+            config.setEnvironments(environments);
```

Comment:
I'd use `EnvironmentsConfig.add` for adding environments.
For removal you can add a `remove` method in `EnvironmentsConfig`, which will search for environment with given name and remove it from list  then removing like this should be safe
```
EnvironmentsConfig environments = config.getEnvironments();
environments.remove(environmentName)
```

Partials won't change much here, because this operation is above all the merging magic. 

Don't call `setEnvironments`, IIRC that is used only from tests, not actual config update operations.